File: kernels/Julia/Compositron/examples/_bench_dbs.jl
using Compositron.Core
using Compositron.Utils
using Compositron.DBS

using Statistics


function work()
    import_bench = @timed Measurement(
        "../../../../testdata/depth-profile_Copper_0000.n42"
    )
    m = import_bench.value
    import_time = import_bench.time * 1e6

    d = m.dbs["OAB"]

    corr_bench = @timed correct_ecal!(d, EcalCorrOrder_First)
    corr_time = corr_bench.time * 1e6

    bg_sub_bench = @timed extract_peak!(d, 30., PeakModel_ErfLinear2Gauss)
    bg_sub_time = bg_sub_bench.time * 1e6

    param = LineshapeParamDefinition(
        "S",
        [PeakArea(KeV(2))],
        [PeakArea(KeV(20))],
        true,
    )

    s_calc_bench = @timed calc_lineshape_param!(d, param, BinIntScheme_Const)
    s_calc_time = s_calc_bench.time * 1e6

    (import_time, corr_time, bg_sub_time, s_calc_time)
end


function print_mean_and_std(name::String, data::AbstractVector)
    m = mean(data)
    s = std(data)

    println("$name: $m µs +/- $s µs")
end


n_iter = 10000

imports = zeros(Float64, n_iter)
corr = zeros(Float64, n_iter)
bg_sub = zeros(Float64, n_iter)
s_calc = zeros(Float64, n_iter)

work()

for i in 1:n_iter
    (import_time, corr_time, bg_sub_time, s_calc_time) = work()

    imports[i] = import_time
    corr[i] = corr_time
    bg_sub[i] = bg_sub_time
    s_calc[i] = s_calc_time
end

print_mean_and_std("Import", imports)
print_mean_and_std("Ecal Corr", corr)
print_mean_and_std("BG Sub", bg_sub)
print_mean_and_std("S Param", s_calc)

File: kernels/Rust/compositron/examples/_bench_dbs.rs
use std::time::Instant;

use compositron::core::Measurement;
use compositron::importers::DataFormat;
use compositron::core::utils::{EcalCorrectionOrder, Unit};
use compositron::dbs::dbspectrum::{Area, LineshapeParamDefinition, PeakModel};


fn print_mean_and_std(name: &str, data: &Vec<u128>) {
    let n = data.len();
    let mean = data.iter().sum::<u128>() as f64 / n as f64;
    let std = (
        data
            .iter()
            .map(|&x| (x as f64 - mean).powi(2))
            .sum::<f64>() / n as f64
    ).sqrt();

    println!("{name}: {mean} ns +/- {std} ns");
}


fn main() -> anyhow::Result<()> {
    let n_iter = 1000;

    let mut imports = Vec::with_capacity(n_iter);
    let mut corr = Vec::with_capacity(n_iter);
    let mut bg_sub = Vec::with_capacity(n_iter);
    let mut s_calc = Vec::with_capacity(n_iter);

    for _ in 0..n_iter {
        let start = Instant::now();
        let mut m = Measurement::from_file(
            "../../../../testdata/depth-profile_Copper_0000.n42",
            DataFormat::SlopeN42
        )?;
        imports.push(start.elapsed().as_nanos());

        let d = m.dbs.get_mut("OAB").unwrap();

        let start = Instant::now();
        d.correct_ecal(EcalCorrectionOrder::First)?;
        corr.push(start.elapsed().as_nanos());

        let start = Instant::now();
        d.extract_peak(30., PeakModel::ErfLinear2Gauss)?;
        bg_sub.push(start.elapsed().as_nanos());

        let param = LineshapeParamDefinition {
            name: "S",
            num: &[Area::Peak { width: Unit::KeV(2.) }],
            denom: &[Area::Peak { width: Unit::KeV(20.) }],
            in_corrected_peak: true,
        };

        let start = Instant::now();
        d.calc_lineshape_param(&param, compositron::dbs::dbspectrum::BinIntegrationScheme::Const)?;
        s_calc.push(start.elapsed().as_nanos());
    }

    print_mean_and_std("Import", &imports);
    print_mean_and_std("Ecal Corr", &corr);
    print_mean_and_std("BG Sub", &bg_sub);
    print_mean_and_std("S Param", &s_calc);

    Ok(())
}
